Count and log consumed energy in Wh

hsteinhaus opened the pull request:
Currently, we do our remaining battery calculations on capacity, i.E. spent mAh. This works well with high-C LiPo batteries, that have a prominent plateau of constant voltage during 70-80% of the flight. 

Modern alternatives with better engergy-to-weight ratio, like 18650 LiIon cells or low-C LiPos, do not have this characteristics. Instead, the voltage is (nearly) constantly decreasing from 4.2V before takeoff to 2.7V at during landing. Periods of heavy load might change the battery temperature, causing arbitrary steps in the voltage. With the voltage going down and the average current going up by 50%, the remaining shown battery percentage is far too optimistic.

![18650](https://user-images.githubusercontent.com/3057137/34245717-8a5f6786-e62b-11e7-8091-4d2445d6f026.png)

As a first step, this PR adds calculation and logging of the consumed ernergy in mWh (as an addition to the capacity in mAh).

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -119,6 +120,10 @@ class AP_BattMonitor
     float current_total_mah(uint8_t instance) const;
     float current_total_mah() const { return current_total_mah(AP_BATT_PRIMARY_INSTANCE); }
 
+    /// total_mwh - returns total energy drawn since start-up in amp-hours
```
"in milliWatt-hours" please

amilcarlucas commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -1106,10 +1107,10 @@ struct PACKED log_DSTL {
-#define CURR_MULTS  "F??????"
+#define CURR_LABELS "TimeUS,Volt,VoltR,Curr,CurrTot,EnrgTot,Temp,Res"
+#define CURR_FMT    "Qfffffcf"
+#define CURR_UNITS  "sv?A??Ow"
```
Please add a proper unit. Adding a ? is NOT a solution.

amilcarlucas replied:
see https://github.com/ArduPilot/ardupilot/blob/master/libraries/DataFlash/LogStructure.h line 74

Pedals2Paddles commented:
Yes yes yes! The next step after this is Watts capacity and watts remaining, which is far more accurate than the current implementation of MAH capacity and remaining!

Would it be more accurate to use the battery monitor's _estimated resting voltage_ rather than the voltage under load?  That would prevent the voltage sags under load from affecting the mathematics.  

I have an open WIP PR / Project on energy used/remaining.  Part of that is calculating watts used, and I could never get it to work properly. Not because it can't be done, but because I don't know what I'm doing :).  So if this patch you've done works, it makes my life much easier!!

hsteinhaus commented:
@amilcarlucas Thanks for finding those nasty c'n'p errors - fixed. However, I am a bit unsure about the best unit. May be switching to Wh (instead of mWh) might be a better choice? How is the multiplier system working when more than one mult is needed? Define a new one?

@Pedals2Paddles 
Not too much magic included. The current code logs mAh, I just multiply those with latest voltage reading and sum up. The result I got seems reasonable when comparing to cell datasheets. But you definitely need a decent power module - the Chinese stuff usually returns random number for current. ACSP4 is working well here.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -71,6 +71,7 @@ class AP_BattMonitor
         float       voltage;            // voltage in volts
         float       current_amps;       // current in amperes
         float       current_total_mah;  // total current draw since start-up
+        float       total_mwh;           // total energy draw since start-up
```
Not to nitpick, but could this be named `consumed_mwh` or similar? Once the concept of a power sources max energy is also present this becomes a naming problem.

hsteinhaus commented:
> and add it to the BATTERY_STATUS message please don't ask me why it's in HectoJoules there
Very good idea - will add that.

> without giving any new data (the logging currently gives you all the data to recreate this in post flight analysis with only logging drops being the concern).

This also applies to current_total_mah. Unfortunately, recreating these data during analysis is rather hard (requires integration), which is only supported by the most sophisticated log processing tools.

Regarding units: in between I favour changing the unit to just plain Wh, as I do not see any use case for the precision of "milli" Watt hours. Any objections against changing that?

WickedShell commented:
Nope, I'd be happier with Wh in a float personally so I say go for it.

Pedals2Paddles commented:
> This also applies to current_total_mah. Unfortunately, recreating these data during analysis is rather hard (requires integration), and is only supported by the most sophisticated log processing tools.

Current total is used in many places on-board and on the GCS for user information.  So it has actual uses now.  Total consumed Wh would not have an _immediate_ use in operation.  However, I think it is valuable information we've been lacking.  And having it allows for other things to build upon it.  There's a million things in the dataflash logs that rarely or never get used for anything, but it has valuable purpose for something sometime.

In other words, as long as this isn't making a problematic dent in overhead, I think it's a good idea.

WickedShell commented:
@Pedals2Paddles I think @hsteinhaus was responding to my desire to rename the variable he used from `total_mwh` to `consumed_mwh` or similair and was pointing out we were already calling it total for all the other current monitoring uses.

Pedals2Paddles commented:
oh.  Well to that point, `consumed_` would actually be a better variable name for all that stuff in the battery monitor libraries.  Saying `total_` is kind of ambigious.  Total what... used? consumed? remaining?  But that's probably for another PR.

amilcarlucas commented:
I would like to see a second PR with the renaming of total_ to consumed_

dgrat commented:
I see no reason not to add this. @WickedShell  The additional processing power is neglectable. @Pedals2Paddles I think you missed that this would break energy calculation. @hsteinhaus Can you show us a diagram with mWh. Should be much more linear at the end of the lipo charge in comparison to the mAh.

hsteinhaus commented:
Here is a plot of energy vs. capacity. The flight was a very steady survey mission, 90° across the wind (120 acres), plus some hovering after work was finished:

![energy_current](https://user-images.githubusercontent.com/3057137/34917268-70cf5002-f944-11e7-904b-0e555a4f8272.png)

dgrat commented:
@hsteinhaus 
as expected. i like it as it is.

WickedShell commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -84,6 +84,7 @@ const struct UnitStructure log_Units[] = {
     { 'G', "Gauss" },         // Gauss is not an SI unit, but 1 tesla = 10000 gauss so a simple replacement is not possible here
     { 'h', "degheading" },    // 0.? to 359.?
     { 'i', "A.s" },           // Ampere second
+    { 'J', "W.s" },           // Joule (Watt second)
```
Why is this not Wh? That seems like the most common unit for considering power consumption?

hsteinhaus replied:
Unit is "Watt seconds"  to keep consistent with Ampere seconds for capacity. 

 "J" is only used as MultiplierStructure.ID, which is, by pure coincidence, similar to Joule, which is, by another coincidence, one Watt times one second ;)

WickedShell replied:
Sorry, I'm not quite following yet! I don't even see the Ampere second being used anywhere? CurTot is currently not tagged with a unit. If you think it's a better unit for log analysis that's fine, but given that we are just dumping a Wh number directly already as a float and it's a reasonable unit to accumulate energy consumption I'm curious why we would want to request GCS's scale it, when not scaling will leave us with a valid result, and graph correctly in those that don't scale.

WickedShell commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -197,7 +197,7 @@ void GCS_MAVLINK::send_battery_status(const AP_BattMonitor &battery, const uint8
                                     battery.has_current(instance) ? battery.current_amps(instance) * 100 : -1, // current
                                     battery.has_current(instance) ? battery.current_total_mah(instance) : -1, // total current
-                                    -1, // joules used
+                                    battery.has_current(instance) ? battery.consumed_wh(instance) * 36 : -1, // consumed energy in hJ (hecto-Joules)
```
While checking `battery.has_current()` seems like a valid approach this does not consider the case of the smart batteries which aren't implementing the Wh counting. You could either implement it in the smart battery drivers (which is probably more work then you want to do I'm guessing), or the approach of either checking that consumed_wh(instanace) is not 0 would also work. The other option you have is to just set the default value of the consumed_Wh in the state to -1 which would result in the analog batteries working easily, and provide the correct MAVLink message for the smart batteries. (I would recommend the latter option as it's the easiest/most robust)

WickedShell commented:
@dgrat I wasn't intending to imply it was something that shouldn't be merged. I've also realized there are actually 2 other very good reasons for bringing this in:
1. Log file integration only works if you log while disarmed (which I do by default, but most people don't do), and if you log a lot of data (which I also do) seeing messages get dropped is also quite common, which would break integration issues.
2. Some smart batteries report in Wh instead of Ah (it's completely acceptable per the protocol to do this, but we currently just assume all smart batteries are Ah, so this is a nice base for future enhancement.

hsteinhaus commented:
I have added a method has_consumed_energy() to the battery backend class. I think that way every implementation is free to chose if consumed energy is supported or not. It defauls to false, and I have implemented it for AP_BattMonitorAnalog only so far.

peterbarker commented:
On Tue, 16 Jan 2018, WickedShell wrote:

> Why is this not Wh? That seems like the most common unit for considering power consumption?

These are all base SI units where possible.  The GCS can interpret if it 
feels the need.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -299,6 +310,15 @@ float AP_BattMonitor::current_total_mah(uint8_t instance) const {
+/// consumed_wh - returns energy consumed since start-up in watt-hours
+float AP_BattMonitor::consumed_wh(uint8_t instance) const {
+    if (instance < _num_instances) {
+        return _BattMonitor_STATE(instance).consumed_wh;
```
`_BattMonitor_STATE()` macro no longer exists in master. `state[instance]` is the appropriate substitution.

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -71,6 +71,7 @@ class AP_BattMonitor
         float       voltage;            // voltage in volts
         float       current_amps;       // current in amperes
         float       current_total_mah;  // total current draw since start-up
```
Please rename this from `current_total_mah` to `consumed_mah`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -119,6 +124,10 @@ class AP_BattMonitor
     float current_total_mah(uint8_t instance) const;
```
Please rename this from `current_total_mah` to `consumed_mah`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -119,6 +124,10 @@ class AP_BattMonitor
     float current_total_mah(uint8_t instance) const;
     float current_total_mah() const { return current_total_mah(AP_BATT_PRIMARY_INSTANCE); }
```
Please rename this from `current_total_mah` to `consumed_mah`

amilcarlucas commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -664,6 +666,7 @@ struct PACKED log_Current {
     float    voltage_resting;
     float    current_amps;
     float    current_total;
```
Please rename this from `current_total` to `consumed_mah`

WickedShell commented:
@amilcarlucas That's all unneeded scope creep. That can be done in a separate PR, but there is no reason for @hsteinhaus to chase that in this PR. All that needs to happen is a rebase and the usage of the removed macro needs to be fixed.

amilcarlucas commented:
An extra single commit with a couple of renames would IMHO be a cherry on top.

hsteinhaus commented:
@WickedShell 
Done that in my working copy already, but have run into some trouble with newer gcc (PX4NuttX no longer builds with them). Will push as soon as this is sorted out.

WickedShell commented:
Merged, thanks!
